From f763867e9a384d6cc37e1374161dc906ea1dd9e0 Mon Sep 17 00:00:00 2001
From: Corentin Oparowski <corentin.oparowski@vates.tech>
Date: Sat, 19 Sep 2015 08:09:42 +0200
Subject: [PATCH 24/36] Use MoK variable fallback

Shim is supposed to create an EFI table containing the MoK configuration which
the kernel accesses via the efi_mokvar_* functions. In some cases (possibly due
to Xen), this table is not available. The configuration is also made available
via EFI variables and the kernel has fallback paths that use the variables if
the table is not available. Fall back to using the EFI variable in this case
too.

With this change, MoK keys correctly get added to the machine keying which is
trusted and can be used to verify modules (rather than the untrusted platform
keyring).

Signed-off-by: Ross Lagerwall <ross.lagerwall@citrix.com>
---
 security/integrity/platform_certs/machine_keyring.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/security/integrity/platform_certs/machine_keyring.c b/security/integrity/platform_certs/machine_keyring.c
index a401640a63cd1..2084bdfd3f3f3 100644
--- a/security/integrity/platform_certs/machine_keyring.c
+++ b/security/integrity/platform_certs/machine_keyring.c
@@ -52,13 +52,18 @@ void __init add_to_machine_keyring(const char *source, const void *data, size_t
 static __init bool uefi_check_trust_mok_keys(void)
 {
 	struct efi_mokvar_table_entry *mokvar_entry;
+	efi_status_t status;
+	unsigned int val = 0;
+	unsigned long size = sizeof(val);
+	efi_guid_t guid = EFI_SHIM_LOCK_GUID;
 
 	mokvar_entry = efi_mokvar_entry_find("MokListTrustedRT");
 
 	if (mokvar_entry)
 		return true;
 
-	return false;
+	status = efi.get_variable(L"MokListTrustedRT", &guid, NULL, &size, &val);
+	return status == EFI_SUCCESS;
 }
 
 static bool __init trust_moklist(void)
-- 
2.47.3

